server/base/src/main/java/org/apache/accumulo/server/master/balancer/HostRegexTableLoadBalancer.java

```diff
@@ -0,0 +1,314 @@
+ * <b>table.custom.balancer.host.regex.is.ip=true</b>
+ *
+ */
+public class HostRegexTableLoadBalancer extends TableLoadBalancer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HostRegexTableLoadBalancer.class);
+  public static final String HOST_BALANCER_PREFIX = Property.TABLE_ARBITRARY_PROP_PREFIX.getKey() + "balancer.host.regex.";
+  public static final String HOST_BALANCER_OOB_CHECK = Property.TABLE_ARBITRARY_PROP_PREFIX.getKey() + "balancer.host.regex.oob.period";
+  private static final String HOST_BALANCER_OOB_DEFAULT = "5m";
+  public static final String HOST_BALANCER_POOL_RECHECK_KEY = Property.TABLE_ARBITRARY_PROP_PREFIX.getKey() + "balancer.host.regex.pool.check";
+  private static final String HOST_BALANCER_POOL_RECHECK_DEFAULT = "5m";
+  public static final String HOST_BALANCER_REGEX_USING_IPS = Property.TABLE_ARBITRARY_PROP_PREFIX.getKey() + "balancer.host.regex.is.ip";
+  protected static final String DEFAULT_POOL = "HostTableLoadBalancer.ALL";
+
+  protected long oobCheckMillis = AccumuloConfiguration.getTimeInMillis(HOST_BALANCER_OOB_DEFAULT);
+  protected long poolRecheckMillis = AccumuloConfiguration.getTimeInMillis(HOST_BALANCER_POOL_RECHECK_DEFAULT);
+
+  private Map<String,String> tableIdToTableName = null;
+  private Map<String,Pattern> poolNameToRegexPattern = null;
+  private long lastOOBCheck = System.currentTimeMillis();
+  private long lastPoolRecheck = 0;
+  private boolean isIpBasedRegex = false;
+  private Map<String,SortedMap<TServerInstance,TabletServerStatus>> pools = new HashMap<String,SortedMap<TServerInstance,TabletServerStatus>>();
+
+  /**
+   * Group the set of current tservers by pool name. Tservers that don't match a regex are put into a default ppol.
+   *
+   * @param current
+   *          map of current tservers
+   * @return current servers grouped by pool name, if not a match it is put into a default pool.
+   */
+  protected synchronized Map<String,SortedMap<TServerInstance,TabletServerStatus>> splitCurrentByRegex(SortedMap<TServerInstance,TabletServerStatus> current) {
+    if ((System.currentTimeMillis() - lastPoolRecheck) > poolRecheckMillis) {
+      Map<String,SortedMap<TServerInstance,TabletServerStatus>> newPools = new HashMap<String,SortedMap<TServerInstance,TabletServerStatus>>();
+      for (Entry<TServerInstance,TabletServerStatus> e : current.entrySet()) {
+        String tableName = getPoolNameForHost(e.getKey().host());
+        if (!newPools.containsKey(tableName)) {
+          newPools.put(tableName, new TreeMap<TServerInstance,TabletServerStatus>());
+        }
+        newPools.get(tableName).put(e.getKey(), e.getValue());
+      }
+      // Ensure that no host is in more than one pool
+      // TODO: I'm not sure that I need to check for disjoint as the call to getPoolNameForHost checks for more than one match
+      boolean error = false;
+      for (SortedMap<TServerInstance,TabletServerStatus> s1 : newPools.values()) {
```

Comment:
This nested loop is doing a lot more disjoint checks than need to be done. You could change the inner loop to iterate over a tailMap starting from the current outer loop value to get rid of the dups.
That being said, I can also think of cases where you might actually want disjoint pools. For example, one pool could be for "archive" data which is large and not accessed very often, so you want to spread it across all tservers. Then other pools, for hot tables would be on subsets of the tservers. To support this would mean some re-architecting of the code since the check would need to be get pool for host and table name.